minplascalc/species.py

```diff
@@ -0,0 +1,306 @@
+
+
+class BaseSpecies:
+    def partitionfunction_total(self, V, T, dE):
+        return (V * self.partitionfunction_translational(T)
+                * self.partitionfunction_internal(T, dE))
+
+    def partitionfunction_translational(self, T):
+        return ((2 * numpy.pi * self.molarmass * constants.Boltzmann * T)
+                / (constants.Avogadro * constants.Planck ** 2)) ** 1.5
+
+    def partitionfunction_internal(self, T, dE):
+        raise NotImplementedError
+
+    def internal_energy(self, T, dE):
+        raise NotImplementedError
+
+
+class Species(BaseSpecies):
+    def __init__(self, name, stoichiometry, molarmass, chargenumber):
+        """Base class for heavy particles. Either single monatomic or diatomic 
+        chemical species in the plasma, eg O2 or Si+
+
+        Parameters
+        ----------
+        name : string
+            A unique identifier for the species.
+        stoichiometry : dictionary
+            Dictionary describing the elemental stoichiometry of the species.
+        molarmass : float
+            Molar mass of the species in kg/mol.
+        chargenumber : int
+            Charge on the species (in integer units of the fundamental charge).
+        """
+        self.name = name
+        self.stoichiometry = deepcopy(stoichiometry)
+        self.molarmass = molarmass
+        self.chargenumber = chargenumber
+        if self.chargenumber < 0:
+            raise ValueError('Negatively charged ions not implemented.')
+
+
+class Monatomic(Species):
+    def __init__(self, name, stoichiometry, molarmass, chargenumber, 
+                 ionisationenergy, energylevels, sources):
+        """Class for monatomic plasma species (single atoms and ions).
+    
+        Parameters
+        ----------
+        name : string
+            A unique identifier for the species.
+        stoichiometry : dictionary
+            Dictionary describing the elemental stoichiometry of the species 
+            (e.g. {'O': 1} for O or O+).
+        molarmass : float
+            Molar mass of the species in kg/mol.
+        chargenumber : int
+            Charge on the species (in integer units of the fundamental charge).
+        ionisationenergy : float
+            Ionisation energy of the species in J.
+        energylevels : list of length-2 lists
+            Atomic energy level data - each entry in the list contains a pair of
+            values giving the level's quantum number and its energy
+            respectively, with energy in J.
+        sources : list of str
+            Each entry represents a reference from which the data was
+            obtained.
+        """
+        super().__init__(name, stoichiometry, molarmass, chargenumber)
+        
+        self.ionisationenergy = ionisationenergy
+        self.energylevels = deepcopy(energylevels)
+        self.sources = deepcopy(sources)
+
+    def __repr__(self):
+        return ('MonatomicSpecies(name=' + self.name + ','
```

Comment:
This is a good motivation for using `self.__class__.__name__` instead of literal strings.

Reply:
Good spot :-/, I'll implement.

Reply:
Good spot... I'll implement.